File: tests/typescript-zod/recursive_union/expected.ts
import { z } from "zod";

export const BaseFieldSchema = z.object({
  name: z.string(),
  required: z.boolean(),
  type: z.string(),
});
export type BaseField = z.infer<typeof BaseFieldSchema>;

export const TextFieldSchema = z.object({
  type: z.literal("text"),
  name: z.string(),
  required: z.boolean(),
});
export type TextField = z.infer<typeof TextFieldSchema>;

export const ObjectFieldSchema = z.object({
  type: z.literal("object"),
  get fields(): z.ZodOptional<z.ZodArray<z.ZodType>> { return z.array(FieldSchemaSchema).optional(); },
  name: z.string(),
  required: z.boolean(),
});
export interface ObjectField {
  type: 'object';
  fields?: FieldSchema[];
  name: string;
  required: boolean;
}

export const ArrayFieldSchema = z.object({
  type: z.literal("array"),
  get fields(): z.ZodOptional<z.ZodArray<z.ZodType>> { return z.array(FieldSchemaSchema).optional(); },
  name: z.string(),
  required: z.boolean(),
});
export interface ArrayField {
  type: 'array';
  fields?: FieldSchema[];
  name: string;
  required: boolean;
}

export const FieldSchemaSchema = z.union([
  TextFieldSchema,
  ObjectFieldSchema,
  ArrayFieldSchema,
]);
export type FieldSchema = TextField | ObjectField | ArrayField;

export const RootSchema = z.object({
  fields: z.array(FieldSchemaSchema),
});
export type Root = z.infer<typeof RootSchema>;

File: tests/python/allof_inheritance/expected.py
from __future__ import annotations

from typing import Annotated, List, Literal, Union
from pydantic import BaseModel, ConfigDict, Field




class BaseField(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    required: bool
    type: str


class TextField(BaseField):
    type: Literal["text"]


class ObjectField(BaseField):
    type: Literal["object"]
    fields: List[FieldSchema] | None = None


class ArrayField(BaseField):
    type: Literal["array"]
    fields: List[FieldSchema] | None = None

FieldSchema = Annotated[
    Union[TextField, ObjectField, ArrayField],
    Field(discriminator="type"),
]



class Root(BaseModel):
    model_config = ConfigDict(extra="forbid")

    fields: List[FieldSchema]

